modules/core/src/main/scala/zio/kafka/streams/TopologySdk.scala

```diff
@@ -0,0 +1,69 @@
+package zio.kafka.streams
+
+import org.apache.kafka.common.serialization.Serde
+import org.apache.kafka.streams.Topology
+import org.apache.kafka.streams.scala.StreamsBuilder
+import org.apache.kafka.streams.scala.kstream.{ Consumed, KStream, Produced }
+import zio.prelude.State
+import zio.prelude.fx.ZPure
+
+trait TopologySdk {
```

Comment:
What are the benefits of passing always `(stream: KStream[K, V])` in the signatures and extending `TopologySdk` vs the "wrapper approach"? Would there be any benefits in wrapping the state instead and keeping everything in the same class?

I'm just thinking aloud, I'm not familiar at all with the `zio` and `zio-preloude` codebases, so I might have completely misunderstood the internals... but I've seen an interesting pattern that we might be able to reuse here too!
Please correct me if I'm saying something stupid, but [this](https://github.com/zio/zio/blob/master/core/shared/src/main/scala/zio/ZIO.scala#L4249) and [this](https://github.com/zio/zio-prelude/blob/master/core/shared/src/main/scala/zio/prelude/fx/ZPure.scala#L907) look like a kind of state machine. Maybe we can do something similar to pass between the 3 main phases (`builder -> source`, `kstream/ktable/...`, `builder => sink`). Would this help with the problem of the signature of the `sink`? Thoughts?

Anyway, before changing the current impl, there are other few considerations to keep in mind.

Should we move away from the official Kakfa Scala lib and wrap directly the Java API? I don't really see a huge benefit in 3 layers Java/Scala/ZIO
* this might help and give us full control when Scala 3 comes around, otherwise we would stick to the scala version of the official lib
* most of the methods with `final Named named` are not directly accessible, it might sounds silly, but if you wanna use something like [eisner](https://github.com/laserdisc-io/eisner) you can't rename the nodes easily
* get rid of these implicits `org.apache.kafka.streams.scala.ImplicitConversions/Serdes` for instance and probably write more "fp friendly" signatures

Also, I would like to come up with a nice way of dealing with [effects](https://github.com/compstak/KafkaStreams4s/issues/79) in general, invoking an endpoint is something that you might don't wanna do for other reasons, but logging, metrics probably yes... If we find a nice way to integrate directly `zio.stream.ZStream` for example, we might open the doors to interesting things...

Reply:
I don't get the first question. You don't pass the stream just because you want to do `stream.doStuff`, you pass it because there might be many streams and you have to decide which one you want to use. Like, after a branch you have two or more streams, if you want to sink the second, how do you write it? Did I understand your question? Can you elaborate on "the problem of the signature of the `sink`"? I don't understand what you are referring to.

All of the three points are good, I just think it will need a bit of work, and we can expose all methods to give name to nodes and get rid of the implicits as well.

About the effects, I think we are good. What would be the problem of integrating effects in `mapValues` for example?

Reply:
> What would be the problem of integrating effects in `mapValues` for example?

Ignore that it's written with cats, but honestly, I didn't find a nice way to do something similar, have you tried?
```scala
// O could be (KO, VO) or Unit depending on the use case
def runMyF[O](f: (K, V) => F[O])(
  implicit F: Sync/Async[F]
): F[???] =
```

About the first question, in most of the cases you will have to pass around always `stream: KStream[K, V]`, so what I was suggesting is to differentiate the cases where you are changing the types in different classes for example. As I said, it was just a thought and it might be over complicating. I'm was saying to combine the 2 approaches and see if there is any benefits